server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +346,298 @@ public DirectoryReader getDirectoryReader() {
+            this.cancellable.checkDirReaderCancelled();
+            return in.next();
+        }
+    }
+
+    /**
+     * Wrapper class for {@link PointValues} that checks for query cancellation or timeout.
+     */
+    private static class ExitablePointValues extends PointValues {
+
+        private final PointValues in;
+        private final Cancellable cancellable;
+
+        private ExitablePointValues(PointValues in, Cancellable cancellable) {
+            this.in = in;
+            this.cancellable = cancellable;
+            this.cancellable.checkDirReaderCancelled();
+        }
+
+        @Override
+        public void intersect(IntersectVisitor visitor) throws IOException {
+            cancellable.checkDirReaderCancelled();
+            in.intersect(new ExitableIntersectVisitor(visitor, cancellable));
+        }
+
+        @Override
+        public long estimatePointCount(IntersectVisitor visitor) {
+            cancellable.checkDirReaderCancelled();
+            return in.estimatePointCount(visitor);
+        }
+
+        @Override
+        public byte[] getMinPackedValue() throws IOException {
+            cancellable.checkDirReaderCancelled();
+            return in.getMinPackedValue();
+        }
+
+        @Override
+        public byte[] getMaxPackedValue() throws IOException {
+            cancellable.checkDirReaderCancelled();
+            return in.getMaxPackedValue();
+        }
+
+        @Override
+        public int getNumDimensions() throws IOException {
+            cancellable.checkDirReaderCancelled();
+            return in.getNumDimensions();
+        }
+
+        @Override
+        public int getNumIndexDimensions() throws IOException {
+            cancellable.checkDirReaderCancelled();
+            return in.getNumIndexDimensions();
+        }
+
+        @Override
+        public int getBytesPerDimension() throws IOException {
+            cancellable.checkDirReaderCancelled();
+            return in.getBytesPerDimension();
+        }
+
+        @Override
+        public long size() {
+            cancellable.checkDirReaderCancelled();
+            return in.size();
+        }
+
+        @Override
+        public int getDocCount() {
+            cancellable.checkDirReaderCancelled();
+            return in.getDocCount();
+        }
+    }
+
+    private static class ExitableIntersectVisitor implements PointValues.IntersectVisitor {
+
+        private static final int MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK = 10;
+
+        private final PointValues.IntersectVisitor in;
+        private final Cancellable cancellable;
+        private int calls;
+
+        private ExitableIntersectVisitor(PointValues.IntersectVisitor in, Cancellable cancellable) {
+            this.in = in;
+            this.cancellable = cancellable;
+        }
+
+        private void checkAndThrowWithSampling() {
+            if (calls++ % MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK == 0) {
```

Comment:
If we want to reduce the overhead to a minimum, we could make `MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK` a power of two minus 1 (e.g. `(1 << 4) - 1`, ie. `15`, and use a mask instead of a remainder: `(calls++ & MAX_CALLS_BEFORE_QUERY_TIMEOUT_CHECK) == 0`.

Reply:
Could you please explain why power of two - 1 is better?

Reply:
mostly because masks are typically cheaper than remainders

Reply:
Thank you!